Multithreading in perl (#122)
Clinical-Genomics/development: multithreading-in-perl/README.md

# Multithreading in perl

The threads module in perl can be used to run multiple processes in parallel. This is an example of a script that use multithreading:

    #!/usr/bin/perl
    # Template for multithreading in perl

    use strict;
    use warnings;
    use threads;

>     my $n_threads = 4; # How many processes that will be done in parallel

Reviewer: ```suggestion
    # How many processes that will be done in parallel
    my $MAX_THREADS = 4;
```

>     my @thr = ();

Reviewer: ```suggestion
    my @threads;
```

    # Loop that starts multiple tasks in parallel on separate threads

>     for (my $i = 0; $i < $n_threads; $i++) {
>         $thr[$i] = threads->create({'context' => 'list'}, \&generic_sub, $i);
>     }

Reviewer: ```suggestion
    for my $thread_index (0 .. $MAX_THREADS) {
        $threads[$thread_index] = threads->create({'context' => 'list'}, \&generic_sub, $thread_index);
    }
```

    # To return calculations from threads

>     for (my $i = 0; $i < $n_threads; $i++) {
>         my ($returned_tid, $returned_sum) = $thr[$i]->join();

Reviewer: ```suggestion
    for my $thread_index (0 .. $MAX_THREADS) {
        my ($returned_tid, $returned_sum) = $threads[thread_index]->join();
```

>         print "The thread with thread ID " . $returned_tid . " got the sum " . $returned_sum . "\n";

Reviewer: ```suggestion
        say {STDOUT} qq{The thread with thread ID $returned_tid  got the sum  $returned_sum};
```

Author: will really {STDOUT} work? I do {\*STDOUT} instead

    }

    # Subroutine that is done multiple times in parallel
    sub generic_sub {
        my ($thread_id) = @_;

>         print "Thread " . $thread_id . " started" . "\n";

Reviewer: ```suggestion
        say {STDOUT} qq{Thread $thread_id started};
```

Author: I use {\\*STDOUT} here as well

Reviewer: No, you are right I forgot the "*" It should be {*STDOUT}

Author: {\*STDOUT} or {\\*STDOUT}?

        my $sum = 0;

>         for (my $i = 0; $i < 10000000; $i++) {

Reviewer: Test to redo the loop according to my suggestions above

            $sum = $sum + rand();
        }

>         print "Thread " . $thread_id . " done" . "\n";

Reviewer: And modify the print to a say

Author: After testing the script with `for f in {1..20}; do perl multi_thread.pl; done` I notice that there is an error that occur when say is used: "String found where operator expected". This error does not occur when print is used. This error does not occur every time the script is executed, so I guess there has to be some unlucky timing and a lot of processes try to do something with STDOUT at the same time?
